File: frontend/lib/models/client_profile.dart
// lib/models/client_profile.dart
import 'dart:convert';

class ClientProfile {
  final int? id;
  final int? userId;

  final String? companyName;
  final String? companySize;
  final String? companyWebsite;
  final String? industry;
  final String? companyDescription;
  final String? companyLogo;
  final int? foundedYear;
  final String? companyType;

  final String? tagline;
  final String? bio;
  final String? location;
  final String? country;
  final String? timezone;
  final String? phone;

  final List<String>? preferredSkills;
  final List<String>? hiringFor;
  final String? preferredContractType;
  final double? budgetRangeMin;
  final double? budgetRangeMax;
  final double? avgProjectBudget;

  final double? totalSpent;
  final int? totalProjects;
  final int? activeContracts;
  final int? completedContracts;
  final int? cancelledContracts;
  final int? hireRate;
  final int? avgContractDuration;
  final double? repeatHireRate;

  final double? clientRating;
  final int? totalReviewsReceived;
  final int? paymentVerificationScore;

  final bool? paymentVerified;
  final bool? idVerified;
  final bool? companyVerified;
  final String? verificationLevel;

  final String? linkedin;
  final String? twitter;
  final String? facebook;
  final String? instagram;

  final List<String>? preferredCommunicationMethods;
  final int? responseTimePreference;
  final List<String>? meetingAvailability;

  final List<String>? projectManagementTools;
  final List<String>? teamCollaborationTools;

  final int? profileViews;
  final int? jobsPosted;
  final int? invitationsSent;
  final int? applicationsReceived;

  final String? preferredFreelancerLevel;
  final String? preferredLocationType;

  final List<Map<String, dynamic>>? badges;
  final bool? isTopClient;
  final bool? isFeaturedClient;

  final int? profileStrength;
  final int? profileCompletionPercentage;

  final DateTime? lastProfileUpdate;
  final DateTime? memberSince;

  ClientProfile({
    this.id,
    this.userId,
    this.companyName,
    this.companySize,
    this.companyWebsite,
    this.industry,
    this.companyDescription,
    this.companyLogo,
    this.foundedYear,
    this.companyType,
    this.tagline,
    this.bio,
    this.location,
    this.country,
    this.timezone,
    this.phone,
    this.preferredSkills,
    this.hiringFor,
    this.preferredContractType,
    this.budgetRangeMin,
    this.budgetRangeMax,
    this.avgProjectBudget,
    this.totalSpent,
    this.totalProjects,
    this.activeContracts,
    this.completedContracts,
    this.cancelledContracts,
    this.hireRate,
    this.avgContractDuration,
    this.repeatHireRate,
    this.clientRating,
    this.totalReviewsReceived,
    this.paymentVerificationScore,
    this.paymentVerified,
    this.idVerified,
    this.companyVerified,
    this.verificationLevel,
    this.linkedin,
    this.twitter,
    this.facebook,
    this.instagram,
    this.preferredCommunicationMethods,
    this.responseTimePreference,
    this.meetingAvailability,
    this.projectManagementTools,
    this.teamCollaborationTools,
    this.profileViews,
    this.jobsPosted,
    this.invitationsSent,
    this.applicationsReceived,
    this.preferredFreelancerLevel,
    this.preferredLocationType,
    this.badges,
    this.isTopClient,
    this.isFeaturedClient,
    this.profileStrength,
    this.profileCompletionPercentage,
    this.lastProfileUpdate,
    this.memberSince,
  });

  factory ClientProfile.fromJson(Map<String, dynamic> json) {
    print('📥 ClientProfile.fromJson: $json');

    List<String> parseStringList(dynamic data) {
      if (data == null) return [];
      if (data is List) {
        return data.map((e) => e.toString()).toList();
      }
      if (data is String) {
        try {
          final parsed = jsonDecode(data);
          if (parsed is List) {
            return parsed.map((e) => e.toString()).toList();
          }
        } catch (e) {}
        return data.split(',').map((s) => s.trim()).toList();
      }
      return [];
    }

    List<Map<String, dynamic>> parseMapList(dynamic data) {
      if (data == null) return [];
      if (data is List) {
        return data.map((e) => Map<String, dynamic>.from(e)).toList();
      }
      if (data is String) {
        try {
          final parsed = jsonDecode(data);
          if (parsed is List) {
            return parsed.map((e) => Map<String, dynamic>.from(e)).toList();
          }
        } catch (e) {}
      }
      return [];
    }

    return ClientProfile(
      id: json['id'],
      userId: json['user_id'] ?? json['UserId'],
      companyName: json['company_name'],
      companySize: json['company_size'],
      companyWebsite: json['company_website'],
      industry: json['industry'],
      companyDescription: json['company_description'],
      companyLogo: json['company_logo'],
      foundedYear: json['founded_year'],
      companyType: json['company_type'],
      tagline: json['tagline'],
      bio: json['bio'],
      location: json['location'],
      country: json['country'],
      timezone: json['timezone'],
      phone: json['phone'],
      preferredSkills: parseStringList(json['preferred_skills']),
      hiringFor: parseStringList(json['hiring_for']),
      preferredContractType: json['preferred_contract_type'],
      budgetRangeMin: json['budget_range_min']?.toDouble(),
      budgetRangeMax: json['budget_range_max']?.toDouble(),
      avgProjectBudget: json['avg_project_budget']?.toDouble(),
      totalSpent: json['total_spent']?.toDouble(),
      totalProjects: json['total_projects'],
      activeContracts: json['active_contracts'],
      completedContracts: json['completed_contracts'],
      cancelledContracts: json['cancelled_contracts'],
      hireRate: json['hire_rate'],
      avgContractDuration: json['avg_contract_duration'],
      repeatHireRate: json['repeat_hire_rate']?.toDouble(),
      clientRating: json['client_rating']?.toDouble(),
      totalReviewsReceived: json['total_reviews_received'],
      paymentVerificationScore: json['payment_verification_score'],
      paymentVerified: json['payment_verified'],
      idVerified: json['id_verified'],
      companyVerified: json['company_verified'],
      verificationLevel: json['verification_level'],
      linkedin: json['linkedin'],
      twitter: json['twitter'],
      facebook: json['facebook'],
      instagram: json['instagram'],
      preferredCommunicationMethods: parseStringList(
        json['preferred_communication_methods'],
      ),
      responseTimePreference: json['response_time_preference'],
      meetingAvailability: parseStringList(json['meeting_availability']),
      projectManagementTools: parseStringList(json['project_management_tools']),
      teamCollaborationTools: parseStringList(json['team_collaboration_tools']),
      profileViews: json['profile_views'],
      jobsPosted: json['jobs_posted'],
      invitationsSent: json['invitations_sent'],
      applicationsReceived: json['applications_received'],
      preferredFreelancerLevel: json['preferred_freelancer_level'],
      preferredLocationType: json['preferred_location_type'],
      badges: parseMapList(json['badges']),
      isTopClient: json['is_top_client'],
      isFeaturedClient: json['is_featured_client'],
      profileStrength: json['profile_strength'],
      profileCompletionPercentage: json['profile_completion_percentage'],
      lastProfileUpdate: json['last_profile_update'] != null
          ? DateTime.tryParse(json['last_profile_update'])
          : null,
      memberSince: json['member_since'] != null
          ? DateTime.tryParse(json['member_since'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'company_name': companyName,
      'company_size': companySize,
      'company_website': companyWebsite,
      'industry': industry,
      'company_description': companyDescription,
      'company_logo': companyLogo,
      'founded_year': foundedYear,
      'company_type': companyType,
      'tagline': tagline,
      'bio': bio,
      'location': location,
      'country': country,
      'timezone': timezone,
      'phone': phone,
      'preferred_skills': preferredSkills,
      'hiring_for': hiringFor,
      'preferred_contract_type': preferredContractType,
      'budget_range_min': budgetRangeMin,
      'budget_range_max': budgetRangeMax,
      'avg_project_budget': avgProjectBudget,
      'total_spent': totalSpent,
      'total_projects': totalProjects,
      'active_contracts': activeContracts,
      'completed_contracts': completedContracts,
      'cancelled_contracts': cancelledContracts,
      'hire_rate': hireRate,
      'avg_contract_duration': avgContractDuration,
      'repeat_hire_rate': repeatHireRate,
      'client_rating': clientRating,
      'total_reviews_received': totalReviewsReceived,
      'payment_verification_score': paymentVerificationScore,
      'payment_verified': paymentVerified,
      'id_verified': idVerified,
      'company_verified': companyVerified,
      'verification_level': verificationLevel,
      'linkedin': linkedin,
      'twitter': twitter,
      'facebook': facebook,
      'instagram': instagram,
      'preferred_communication_methods': preferredCommunicationMethods,
      'response_time_preference': responseTimePreference,
      'meeting_availability': meetingAvailability,
      'project_management_tools': projectManagementTools,
      'team_collaboration_tools': teamCollaborationTools,
      'profile_views': profileViews,
      'jobs_posted': jobsPosted,
      'invitations_sent': invitationsSent,
      'applications_received': applicationsReceived,
      'preferred_freelancer_level': preferredFreelancerLevel,
      'preferred_location_type': preferredLocationType,
      'badges': badges,
      'is_top_client': isTopClient,
      'is_featured_client': isFeaturedClient,
      'profile_strength': profileStrength,
      'profile_completion_percentage': profileCompletionPercentage,
      'last_profile_update': lastProfileUpdate?.toIso8601String(),
      'member_since': memberSince?.toIso8601String(),
    };
  }
}

File: backend/src/models/ClientProfile.js
import { DataTypes } from "sequelize";
import { sequelize } from "../config/db.js";

const ClientProfile = sequelize.define("ClientProfile", {
  id: { type: DataTypes.INTEGER, autoIncrement: true, primaryKey: true },

  company_name: { type: DataTypes.STRING, allowNull: true },
  company_size: {
    type: DataTypes.ENUM("1", "2-10", "11-50", "51-200", "201-1000", "1000+"),
    allowNull: true,
  },
  company_website: { type: DataTypes.STRING, allowNull: true },
  industry: { type: DataTypes.STRING, allowNull: true },
  company_description: { type: DataTypes.TEXT, allowNull: true },
  company_logo: { type: DataTypes.STRING, allowNull: true },
  founded_year: { type: DataTypes.INTEGER, allowNull: true },
  company_type: {
    type: DataTypes.ENUM(
      "startup",
      "small_business",
      "medium_business",
      "enterprise",
      "individual",
    ),
    defaultValue: "individual",
  },

  tagline: { type: DataTypes.STRING(160), allowNull: true },
  bio: { type: DataTypes.TEXT, allowNull: true },
  location: { type: DataTypes.STRING, allowNull: true },
  country: { type: DataTypes.STRING, allowNull: true },
  timezone: { type: DataTypes.STRING, allowNull: true },
  phone: { type: DataTypes.STRING, allowNull: true },

  preferred_skills: { type: DataTypes.TEXT, defaultValue: "[]" },
  hiring_for: { type: DataTypes.TEXT, defaultValue: "[]" },
  preferred_contract_type: {
    type: DataTypes.ENUM("hourly", "fixed", "both"),
    defaultValue: "both",
  },
  budget_range_min: { type: DataTypes.DECIMAL(10, 2), allowNull: true },
  budget_range_max: { type: DataTypes.DECIMAL(10, 2), allowNull: true },
  avg_project_budget: { type: DataTypes.DECIMAL(10, 2), defaultValue: 0 },

  total_spent: { type: DataTypes.DECIMAL(10, 2), defaultValue: 0 },
  total_projects: { type: DataTypes.INTEGER, defaultValue: 0 },
  active_contracts: { type: DataTypes.INTEGER, defaultValue: 0 },
  completed_contracts: { type: DataTypes.INTEGER, defaultValue: 0 },
  cancelled_contracts: { type: DataTypes.INTEGER, defaultValue: 0 },
  hire_rate: { type: DataTypes.INTEGER, defaultValue: 0 },
  avg_contract_duration: { type: DataTypes.INTEGER, defaultValue: 0 },
  repeat_hire_rate: { type: DataTypes.FLOAT, defaultValue: 0 },

  client_rating: { type: DataTypes.FLOAT, defaultValue: 0 },
  total_reviews_received: { type: DataTypes.INTEGER, defaultValue: 0 },
  payment_verification_score: { type: DataTypes.INTEGER, defaultValue: 0 },

  payment_verified: { type: DataTypes.BOOLEAN, defaultValue: false },
  id_verified: { type: DataTypes.BOOLEAN, defaultValue: false },
  company_verified: { type: DataTypes.BOOLEAN, defaultValue: false },
  verification_level: {
    type: DataTypes.ENUM("basic", "verified", "trusted", "premium"),
    defaultValue: "basic",
  },

  linkedin: { type: DataTypes.STRING, allowNull: true },
  twitter: { type: DataTypes.STRING, allowNull: true },
  facebook: { type: DataTypes.STRING, allowNull: true },
  instagram: { type: DataTypes.STRING, allowNull: true },

  preferred_communication_methods: { type: DataTypes.TEXT, defaultValue: "[]" },
  response_time_preference: { type: DataTypes.INTEGER, defaultValue: 24 },
  meeting_availability: { type: DataTypes.TEXT, defaultValue: "[]" },

  project_management_tools: { type: DataTypes.TEXT, defaultValue: "[]" },
  team_collaboration_tools: { type: DataTypes.TEXT, defaultValue: "[]" },

  profile_views: { type: DataTypes.INTEGER, defaultValue: 0 },
  jobs_posted: { type: DataTypes.INTEGER, defaultValue: 0 },
  invitations_sent: { type: DataTypes.INTEGER, defaultValue: 0 },
  applications_received: { type: DataTypes.INTEGER, defaultValue: 0 },

  preferred_freelancer_level: {
    type: DataTypes.ENUM("entry", "intermediate", "expert", "any"),
    defaultValue: "any",
  },
  preferred_location_type: {
    type: DataTypes.ENUM("local", "remote", "hybrid", "any"),
    defaultValue: "any",
  },

  badges: { type: DataTypes.TEXT, defaultValue: "[]" },
  is_top_client: { type: DataTypes.BOOLEAN, defaultValue: false },
  is_featured_client: { type: DataTypes.BOOLEAN, defaultValue: false },

  profile_strength: { type: DataTypes.INTEGER, defaultValue: 0 },
  profile_completion_percentage: { type: DataTypes.INTEGER, defaultValue: 0 },

  last_profile_update: { type: DataTypes.DATE, allowNull: true },
  member_since: { type: DataTypes.DATE, allowNull: true },
  verification_document_url: {
    type: DataTypes.STRING,
    allowNull: true,
  },
  client_type: {
    type: DataTypes.ENUM("individual", "small_business", "enterprise"),
    defaultValue: "individual",
  },
  commercial_register_number: {
    type: DataTypes.STRING,
    allowNull: true,
  },
  commercial_register_image: {
    type: DataTypes.STRING,
    allowNull: true,
  },
  tax_number: {
    type: DataTypes.STRING,
    allowNull: true,
  },
});

export default ClientProfile;
